share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -451,3 +453,120 @@ function(sysbuild_cache_set)
 
   set(${VARS_VAR} "${var_new}" CACHE "${var_type}" "${var_help}" FORCE)
 endfunction()
+
+# Usage:
+#   sysbuild_add_dependencies(<CONFIGURE | FLASH> <image> [<image-dependency> ...])
+#
+# This function makes an image depend on other images in the configuration or
+# flashing order. Each image named "<image-dependency>" will be ordered before
+# the image named "<image>".
+#
+# CONFIGURE: Add CMake configuration dependencies. This will determine the order
+#            in which `ExternalZephyrProject_Cmake()` will be called, as well as
+#            the order of images passed to all PRE_CMAKE and POST_CMAKE hooks.
+# FLASH:     Add flashing dependencies. This will determine the order in which
+#            all images will appear in `domains.yaml`, as well as the order of
+#            images passed to all PRE_DOMAINS and POST_DOMAINS hooks.
+#
+# Note: specifying dependencies on non-existent images is allowed. This makes it
+# possible to define relationships between various images in the build system,
+# regardless of whether they are present in the current multiimage build.
+#
+function(sysbuild_add_dependencies dependency_type image)
+  set(valid_dependency_types CONFIGURE FLASH)
+  if(NOT dependency_type IN_LIST valid_dependency_types)
+    list(JOIN valid_dependency_types ", " valid_dependency_types)
+    message(FATAL_ERROR "sysbuild_add_dependencies(...) dependency type "
+                        "${dependency_type} must be one of the following: "
+                        "${valid_dependency_types}"
+    )
+  endif()
+
+  set(property_name sysbuild_deps_${dependency_type}_${image})
+  set_property(GLOBAL APPEND PROPERTY ${property_name} ${ARGN})
```

Comment:
I used global properties to support using `sysbuild_add_dependencies()` on images that haven't been added yet, in which case the target wouldn't exist.

See this related comment chain: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1193611885 I would appreciate your input there.

Reply:
> I used global properties to support using sysbuild_add_dependencies() on images that haven't been added yet, in which case the target wouldn't exist.

Not a good enough reason, cause it also make the rest of the code more complex.
If dependencies are added on targets you can pass the targets directly to your topology search and each target itself will know it's dependencies.

This mean that all code related to loop your global properties and creation of edges are can simply be removed, such as:
https://github.com/zephyrproject-rtos/zephyr/blob/5535dce7e745430f500a8849d40b718ad0821449/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L551-L566

https://github.com/zephyrproject-rtos/zephyr/blob/5535dce7e745430f500a8849d40b718ad0821449/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L491-L500

Reply:
> This mean that all code related to loop your global properties and creation of edges are can simply be removed, such as:

For example, the following function will work directly on target, and thus all other surrounding code can be eliminated.

Note: ported from here: https://github.com/zephyrproject-rtos/zephyr/blob/7457bcf0a2b3d3523b8b3ea971771337f3751f1f/scripts/zephyr_module.py#L603

```
function(topological_sort)
  cmake_parse_arguments(TS "" "RESULT;PROPERTY_NAME" "TARGETS" ${ARGN})
  
  set(dep_targets)
  set(start_targets)
  set(sorted_targets)
  
  foreach(target ${TS_TARGETS})
    get_target_property(${target}_dependencies ${target} ${TS_PROPERTY_NAME})
  
    if(${target}_dependencies)
      list(APPEND dep_targets ${target})
    else()
      list(APPEND start_targets ${target})
    endif()
  endforeach()
  
  while(TRUE)
    list(POP_FRONT start_targets node)
    list(APPEND sorted_targets ${node})
    set(to_remove)
    foreach(target ${dep_targets})
      if("${node}" IN_LIST ${target}_dependencies)
        list(REMOVE_ITEM ${target}_dependencies ${node})
        if(NOT ${target}_dependencies)
          list(APPEND start_targets ${target})
          list(APPEND to_remove ${target})
        endif()
      endif()
    endforeach()
  
    foreach(target ${to_remove})
      list(REMOVE_ITEM dep_targets ${target})
    endforeach()
    if(NOT start_targets)
      break()
    endif()
  endwhile()
  
  if(dep_targets)
    foreach(target ${dep_targets})
      get_target_property(deps ${target} ${TS_PROPERTY_NAME})
      list(JOIN deps " " deps)
      list(APPEND dep_string "${target} depends on: ${deps}")
    endforeach()
    list(JOIN dep_string "\n" dep_string)
    message(FATAL_ERROR "Unmet or cyclic dependencies in modules:\n${dep_string}")
  endif()
  
  set(${TS_RESULT} "${sorted_targets}" PARENT_SCOPE)
endfunction()
```

Function can easily be tested, like this:
```
add_custom_target(A)
add_custom_target(B)
add_custom_target(C)
add_custom_target(D)

set(IMAGES A B C D)
message("Unsorted images: ${IMAGES}")

set_property(TARGET A APPEND PROPERTY CUSTOM_DEPENDS C)
set_property(TARGET B APPEND PROPERTY CUSTOM_DEPENDS A C)
set_property(TARGET C APPEND PROPERTY CUSTOM_DEPENDS D)

topological_sort(TARGETS ${IMAGES} PROPERTY_NAME CUSTOM_DEPENDS RESULT IMAGES)
message("Sorted images  : ${IMAGES}")
message("==========================\n")

# Add cyclic dep.
set_property(TARGET C APPEND PROPERTY CUSTOM_DEPENDS A)
topological_sort(TARGETS ${IMAGES} PROPERTY_NAME CUSTOM_DEPENDS RESULT IMAGES)
message("Resorted images  : ${IMAGES}")
```
Result:
```
Unsorted images: A;B;C;D                                                                                                                                                                                
Sorted images  : D;C;A;B                                                                                                                                                                                
==========================                                                                                                                                                                              
```

And cyclic errors:
```
CMake Error at CMakeLists.txt:54 (message):                                                                                                                                                             
  Unmet or cyclic dependencies in modules:                                                                                                                                                              
                                                                                                                                                                                                        
  C depends on: A                                                                                                                                                                                       
                                                                                                                                                                                                        
  A depends on: C                                                                                                                                                                                       
                                                                                                                                                                                                        
  B depends on: A                                                                                                                                                                                       
Call Stack (most recent call first):                                                                                                                                                                    
  CMakeLists.txt:79 (topological_sort)              
```

The topological sort is very similar to your current proposal, and feel free to rework that one into use targets (or images) instead of using above proposal.

The main point is to avoid all the additional overhead / maintenance of extra global properties, processing of those properties into additional `,`-separated strings and then pass this on to the sort function when everything can be done directly using target properties.

Reply:
Sorry, `zephyr_topological_sort()` needs to stay generic in order to support groups: https://github.com/zephyrproject-rtos/zephyr/pull/58174

Please have a look at how I extended the `sysbuild_images_order()` function in https://github.com/zephyrproject-rtos/zephyr/pull/58174/commits/314fcf8b3fcb4356b78543d9fd75d11020b7a4cf. There, I construct a single dependency graph based on both image-wise and group-wise dependencies. The target-based sorting you wrote would be applicable for image-wise dependencies only.

Thanks for the effort, though... I should have replied sooner.

Reply:
> Sorry, `zephyr_topological_sort()` needs to stay generic in order to support groups: #58174

Groups are handled outside of `topological_sort()` in a `foreach()` here:
https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L680
https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L816

so I don't see the relevance in relation to the groups PR.

Instead, the use of `,`-separated strings just makes it more complex to use the topological sort function, as every caller must now generate a list of such strings.

A core part of CMake are targets and handling of those, so sticking to targets and properties will remove need for construction of `,`-separated strings.

And groups are already defined as targets: https://github.com/zephyrproject-rtos/zephyr/blob/2b791a1eb2dc489001b6bc244504dadd726e750f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#LL433C5-L433C22

Reply:
I decided to copy your target-based toposort and credit you as the author, if you don't mind. Thanks again for the effort!